### ISSUE TITLE
[improvement] Use the GPU for video transcoding

### ISSUE THREAD

**Issue by peter-sc** [xemle/home-gallery#74]
Hello @xemle , thanks for great app !

Because we are using ffmpeg for transcoding video, would it be possible to use the GPU capabilities? It could be added to the configuration file as an option. The default one would be CPU as currently used, or the one supported by his/her GPU.
It would speed up transcoding multiple times.

I know that RPi and similar SBC don't support it, but users using their laptops to process the library would benefit from it.

**Comment by xemle** [comment]
Hi @peter-sc 

Thank you for reaching out and providing an improvement proposal.

Can you describe how the ffmpeg command will change? Basically in the extractor step in the `video` extractor the videos are transcoded to 720p. Some arguments can be seen [here in video.js](https://github.com/xemle/home-gallery/blob/master/packages/extractor/src/extract/video/video.js#L38).

**Comment by peter-sc** [comment]
The parameters depend on the available GPU, OS and supported encoder options. Somebody might use the GPU for decoding&encoding, some just for encoding.

Transcoding speed comparison (both tested with ffmpeg on Windows):

Current home gallery settings, transcoding speed 1.5x:
`ffmpeg -i input.mov -c:v libx264 -r 30 -vf scale=-2:'min(720,ih)',format=yuv420p -preset slow -tune film -profile:v baseline -level 3.0 -maxrate 4000k -bufsize 8000k -movflags +faststart -b:a 128k -f mp4 output.mp4`

Intel Quicksync, speed 5.6x:
`ffmpeg -i input.mov -c:v h264_qsv -vf scale=-2:'min(720,ih)',format=yuv420p -global_quality 25 output.mp4`

Most universal would be to have an option to manually define the ffmpeg parameters, as the possibilities of running in different enviroments are quite high.

First I was thinking to process all the images and videos on the more powerfull laptop, and then fetch them to the small and low performance server. It turned out, Intel GPU isn't supported in docker/wsl2 yet, only Nvidia.
I also tried to run the native windows binary, and fetch the DB to the home gallery instanance running in docker on Linux, here it turned out that on windows the DB uses different path names compared to Linux DB and it couldn't fetch only the new additions, but everything and it wanted to save all the files with windows path syntax.

**Comment by xemle** [comment]
> The parameters depend on the available GPU, OS and supported encoder options. Somebody might use the GPU for decoding&encoding, some just for encoding.
> 
> Transcoding speed comparison (both tested with ffmpeg on Windows):
> 
> Current home gallery settings, transcoding speed 1.5x: `ffmpeg -i input.mov -c:v libx264 -r 30 -vf scale=-2:'min(720,ih)',format=yuv420p -preset slow -tune film -profile:v baseline -level 3.0 -maxrate 4000k -bufsize 8000k -movflags +faststart -b:a 128k -f mp4 output.mp4`
> 
> Intel Quicksync, speed 5.6x: `ffmpeg -i input.mov -c:v h264_qsv -vf scale=-2:'min(720,ih)',format=yuv420p -global_quality 25 output.mp4`
> 
> Most universal would be to have an option to manually define the ffmpeg parameters, as the possibilities of running in different enviroments are quite high.

It would be possible to provide custom ffmpeg args  via env between input arg `-i input.mov` and output arg `-f mp4 output.mp4`. Then you would be able to customize your profile. Would that work for your? E.g. by setting `GALLERY_FFMPEG_ARGS=-c:v h264_qsv -vf scale=-2:'min(720,ih)',format=yuv420p -global_quality 25`. The user must be responsible how to transcode the videos.

The current profile is choosen for maximum mobile compatibility. To have a good balance between mobile experience, low bandwidth and desktop preview the resolution 720p was choosen.

> First I was thinking to process all the images and videos on the more powerfull laptop, and then fetch them to the small and low performance server. It turned out, Intel GPU isn't supported in docker/wsl2 yet, only Nvidia. I also tried to run the native windows binary, and fetch the DB to the home gallery instanance running in docker on Linux, here it turned out that on windows the DB uses different path names compared to Linux DB and it couldn't fetch only the new additions, but everything and it wanted to save all the files with windows path syntax.

Interesting. I am happy that you want to use the fetch feature. The idea of the fetch feature is to be compatible between the systems. So fetching between Windows, Linux and Mac should be possible. Since I work on Linux the testing of Windows is limited. For the backend process is should not matter to be on Windows or on Linux. On the webapp it should be forward slashes like in URLs

As far as I understood the fetching is not working between Windows and LInux, right? Can you open a separate issue for that?

Further, what do you mean by

> but everything and it wanted to save all the files with windows path syntax.

What workflow do you have with Windows paths? To you need them formated in the Web App in the detail view with backslashes? Maybe open a separate issue here, too.

**Comment by peter-sc** [comment]
The idea of being able to configure ffmpeg args between input and output is great. Either as `GALLERY_FFMPEG_ARGS` or via `gallery.config.yml`, any way is good :)

For the fetch use case I'll try to retest it, collect logs and open a new issue if necessary.

**Comment by xemle** [comment]
`GALLERY_FFMPEG_ARGS` is simpler to implement than `gallery.config.yml`. However `gallery.config.yml` can define explicit args and is less error prone. Therefore, I would go for the later.

I need to change a bit to pass the config values of `gallery.config.yml` to the exporter rather than reading CLI arguments as the current state. Please stay tuned, subscribe to this issue or submit a PR for this feature request.

**Comment by xemle** [comment]
Hi @peter-sc 

The new v1.11.0 is released and supports custom ffmpeg args via the `gallery.config.yml`. See [configuration docs](https://docs.home-gallery.org/configuration/index.html#full-configuration):

```
#extractor:
  #video:
  #  previewSize: 720 # used as preview size and preview filename
  #  ext: mp4 # used as target video container and preview filename
  #  scale: -2:'min(720,ih)' # overwrites previewSize. E.g. -2:'min(ih,max(720,min(1080,ih*.5)))' for 720p <= height*0.5 <= 1080p or height, if height < 720
  #  frameRate: 30
  #  maxVideoBitRate: 4000 # in KB
  #  videoEncoder: libx264
  #  preset: slow
  #  profile: baseline
  #  level: '3.0'
  #  addFfmpegArgs: [...] # Additional ffmpeg args
  #  customFfmpegArgs: [...] # custom ffmpeg args for video conversion, replaces all other settings such previewSize, videoEncoder, addFfmpegArgs, ...
```

I guess you can customize now your video workflow and can speedup the video transcoding.

See further changes on v1.11.0 in the [Changelog](https://docs.home-gallery.org/changelog.html)

**Comment by xorbug** [comment]
Hi,
I was looking exactly for this feature so, since it's still open, I hope it's appropriate to comment under this rather than opening a new issue.

First of all thanks for implementing this and in general for the whole project. I tested this feature a bit, but apparently it is broken, since what is allowed currently is either:
- appending additional args to the end of the ffmpeg command line (via `addFfmpegArgs`)
- replacing the the command line altogether with a custom one (via `customFfmpegArgs`).

Now I'm no expert with ffmpeg, but the first doesn't seem suitable because what is needed is also the ability to inject options before the input argument e.g. `-hwaccel` or `-hwaccel_output_format` (as per [ffmpeg docs](https://trac.ffmpeg.org/wiki/HWAccelIntro) and backend-specific docs). And the second might not be too user-friendly as it requires the user to also know the other technicalities and specify a complete command line.

So what about extending the first solution by adding some more flexibility? For simple cases the ffmpeg command line can roughly be abstracted in two blocks: input options + input, and output options + output. So maybe one could leverage [input options](https://github.com/fluent-ffmpeg/node-fluent-ffmpeg#inputoptionsoption-add-custom-input-options) too for this?

Without bothering about hardcoding hardware acceleration options in the config, maybe a generic approach conforming to this structure (and thus allowing more flexility for whatever feature an user might need in future) could then be having two additional arrays in the config (e.g. `AdditionalInputArgs` and `AdditionalOutputArgs`) to be appended to default input and output options? I think a bit like it is now, but also for the input?

Thanks.

**Comment by xemle** [comment]
Hi @xorbug 

thank you for the flowers and reaching out.

IMHO ffmpeg is an awesome tool which comes with a lot of complexity. I tried to enable some customization if desired. My question to your request is: Can we extend `customFfmpegArgs` to allow `-hwaccel` parameters in front of the source instead of adding further features and settings? So that the advance user can customize ffmpeg for its needs. Would it help you that you can set the position of source and output in the `customFfmepgArgs`? 

An example would be

```
extractor:
  video:
    customFfmpegArgs: [-hwaccel, -i, <source>, ...]
```

If `<source>` is given it will be injected there. Otherwise it will be added in the front. The default arguments can be caught by the logs. So setup of `customFfmpegArgs` should be easy enough.

While it is awesome to tweak the last performance out of ffmpeg my assumption for the implementation is: The standard user of a self hosted web gallery has disk space and time. So the last performance optimization is not that important. If the standard user waits 3.5 weeks to encode his whole videos initially instead of 4 weeks it does not really matter. The need for a stable self hosted solution for the assumed user is higher than having a high optimized video encoder.

What is you opinion? What is your need and what are you expect to gain with an optimized hardware acceleration? 

Further, PRs are welcome. The magic happens in [video.js](https://github.com/xemle/home-gallery/blob/master/packages/extractor/src/extract/video/video.js#L20) which uses [fluent-ffmpeg](https://www.npmjs.com/package/fluent-ffmpeg). The [tests](https://github.com/xemle/home-gallery/blob/master/packages/extractor/src/extract/video/video-utils.test.js) for the ffmpeg arguments should give further hints.

Do not hesitate further questions

**Comment by xorbug** [comment]
> Hi @xorbug
> 
> thank you for the flowers and reaching out.
> 
> IMHO ffmpeg is an awesome tool which comes with a lot of complexity. I tried to enable some customization if desired. My question to your request is: Can we extend `customFfmpegArgs` to allow `-hwaccel` parameters in front of the source instead of adding further features and settings? So that the advance user can customize ffmpeg for its needs. Would it help you that you can set the position of source and output in the `customFfmepgArgs`?
> 
> An example would be
> 
> ```
> extractor:
>   video:
>     customFfmpegArgs: [-hwaccel, -i, <source>, ...]
> ```
> 
> If `<source>` is given it will be injected there. Otherwise it will be added in the front. The default arguments can be caught by the logs. So setup of `customFfmpegArgs` should be easy enough.

Hi @xemle,
sure your solution seems fine too. Just I thought it could be more straightforward to make use of fluent-ffmpeg's [`inputOptions()`](https://github.com/fluent-ffmpeg/node-fluent-ffmpeg#inputoptionsoption-add-custom-input-options) at this point (similarly to how `outputOptions()` is [already used](https://github.com/xemle/home-gallery/blob/d6f71e92bb7567e077f9ec88260c9187fbcafe4a/packages/extractor/src/extract/video/video.js#L38) in the current implementation). So just having an array for custom inputs and one for custom outputs in the config, and then building the command line appending these arrays to the default input/output options (just like it is now for outputs, but for inputs too). But honestly I don't really have any further insight into the code, so was just the first idea I got and don't know how feasible that is.

> While it is awesome to tweak the last performance out of ffmpeg my assumption for the implementation is: The standard user of a self hosted web gallery has disk space and time. So the last performance optimization is not that important. If the standard user waits 3.5 weeks to encode his whole videos initially instead of 4 weeks it does not really matter. The need for a stable self hosted solution for the assumed user is higher than having a high optimized video encoder.
> 
> What is you opinion? What is your need and what are you expect to gain with an optimized hardware acceleration?

Sure i agree on this. But it's also true that the amount of performance gain that one can get should be totally worth it (many times faster, as @peter-sc showed for example with almost 4x improvement), and in general puts less load on the cpu, which might be running other services too, so it's always a good thing... the usual benefits of hardware acceleration basically, especially if it comes (almost) for free (i.e. adding support in code to specify custom parameters). Then of course it's up to the users if the defaults are not ok and they want to take advantage of this. Also I think that once the feature is set up it will be very flexible for any other custom option a user might want to add to ffmpeg.

> Further, PRs are welcome. The magic happens in [video.js](https://github.com/xemle/home-gallery/blob/master/packages/extractor/src/extract/video/video.js#L20) which uses [fluent-ffmpeg](https://www.npmjs.com/package/fluent-ffmpeg). The [tests](https://github.com/xemle/home-gallery/blob/master/packages/extractor/src/extract/video/video-utils.test.js) for the ffmpeg arguments should give further hints.

Yes I would really like to contribute but unfortunately I'm not exactly familiar with node and would take me a bit to get started.

**Comment by xemle** [comment]
Hi @xorbug 

I am fine with `inputOptions()` from fluent-ffmpeg while I would prefer just using pimping `customFfmpegArgs`.

`customFfmpegArgs` will reduce the option complexity for the video extractor instead of introducing further options `addInputArgs` and `addOutputArgs` which might not fit needs from someone else. As I said the logs prints all arguments and it should be simple enough to create `customFfmpegArgs` out of it. IMHO an advanced documentation for different hardware acceleration would be more helpful here than defining additional options like `addInputArgs`. The advance documentation can be added to the [doc repo](https://github.com/xemle/home-gallery-doc).

Would you mind to open an PR for it? That would be great! You can than decide how to implement your solution ;-)

Thank you for your valuable time and input.

**Comment by peter-sc** [comment]
Hi @xemle , 

I'm testing the new `customFfmpegArgs` option, but as @xorbug mentioned, it doesn't seem to work.

My settings are:
`   customFfmpegArgs: [-c:v, h264_qsv, -vf scale=-2:'min(720,ih)',format=yuv420p, -global_quality 25]`

in log file it seems it's ignored:
```
{"level":20,"time":"2023-10-17T19:40:46.710Z","pid":1788,"hostname":"nb","module":"extractor.video","levelName":"debug","msg":"Start video conversion via ffmpeg command: ffmpeg -i c:\\Users\\user\\2023\\04\\2023-03-03_195201_2504.mov -y -y -c:v libx264 -c:a aac -r 30 -vf scale=-2:'min(720,ih)',format=yuv420p -preset slow -tune film -profile:v baseline -level:v 3.0 -maxrate 4000k -bufsize 8000k -movflags +faststart -b:a 128k -f mp4 .cache\\home-gallery\\storage\\d0\\6a\\2b937efe4b611859719dbc8e43e6f9192387-video-preview-720.mp4.tmp"}
```

I'm testing on windows with `home-gallery-1.13.2-win-x64.exe`.

Do I need to include the `-i <source>` and `-o <?>` as well ?

**Comment by xemle** [comment]
@peter-sc thank you for reaching out. I can reproduce the bug and `customFffmpegArgs` are not getting populated as it should be. Sorry for that. I will fix it in the next few days

**Comment by xemle** [comment]
Two findings

- `customFffmpegArgs` is populated if the configuration is found automatically. `./gallery -c gallery.custom.config.yml run import` is currently broken for the run import stand alone job. This will be fixed in a few days. `./gallery.js -c gallery.custom.config.yml run server` works with `customFffmpegArgs`.

- `customFfmpegArgs: [-c:v, h264_qsv, -vf scale=-2:'min(720,ih)',format=yuv420p, -global_quality 25]` is invalid in two ways: 
  - it is not a valid argument array and should be `customFfmpegArgs: [-c:v, h264_qsv, -vf, "scale=-2:'min(720,ih)',format=yuv420p", -global_quality, 25]` so the final aguments are
    - -c:v
    - h264_qsv
    - -vf
    - scale=-2:'min(720,ih)',format=yuv420p
    - -global_quality
    - 25
  - These arguments seems to be broken on my test video conversion and `ffmpeg` fails with return code error 1.

**Comment by peter-sc** [comment]
@xemle thank you for looking into it, the correct arguments which work for me now are:

`customFfmpegArgs: [-c:v, h264_qsv, -vf, "scale=-2:'min(720,ih)',format=yuv420p", -global_quality, 25, -f, mp4]`

Because the video is saved into a tmp file first, the `-f mp4` needs to be present.

**Comment by xemle** [comment]
@peter-sc can you validate which ffmpeg arguments you need by calling ffmpeg directly?

Please test the call with the provided `ffmpeg.exe`. [home-gallery-latest-win-x64.tar.gz](https://dl.home-gallery.org/dist/latest/home-gallery-latest-win-x64.tar.gz) contains these files as `home-gallery-1.13.2-win-x64.exe` but without the self extracting feature. Within the tarball you can access the used `ffmpeg.exe` in `home-gallery/node_modules/@ffmpeg-installer/win32-x64/`.

Further I have the feeling that the ffmpeg's source argument `-i` can be set before a specific video encoder `-c:v h264_qsv` so there is no need to place the input after the codec argument. Can you verify that since I do not have an specific hardware accelerated encoder on my machine.

**Comment by peter-sc** [comment]
This works for me, when calling ffmpeg.exe directly (I needed to add the `" "` for the video filter:

`.\ffmpeg.exe -i 2023-05-27_161517.MOV -y -c:v h264_qsv -vf "scale=-2:'min(720,ih)',format=yuv420p" -global_quality 25 -f mp4 output.mp4`

From my testing, the `-i` argument must be set before the video encoder, otherwise I got errors like:
`Option vf (set video filters) cannot be applied to input url 2023-05-27_161517.MOV -- you are trying to apply an input option to an output file or vice versa. Move this option before the file it belongs to.`

**Comment by xemle** [comment]
Thank you @peter-sc for your example using `customFfmpegArgs`.

Closing this ticket as solved.

**Comment by xemle** [comment]
> Because the video is saved into a tmp file first, the `-f mp4` needs to be present.

@peter-sc In the released 1.14.0 version, `-f mp4` can be omitted and is set automatically.
